File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        return this.solution0(head)
        // return solution1(head)
    }


    solution0(head) {
        let prev = null
        let curr = head
         
         while (curr) {
            const next = curr.next;
            curr.next = prev;
            [prev, curr] = [curr, next]
         }

         return prev
    }

    solution1(head) {

    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# time - O(max(n,m)), space - O(max(n,m))
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        sum = 0
        dummy = ListNode(0)
        cur = dummy
        while l1 or l2:
            sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            carry = sum//10
            digit = sum%10
            cur.next = ListNode(digit)
            cur = cur.next
            if l1: l1 = l1.next
            if l2: l2 = l2.next
        if (carry > 0): cur.next = ListNode(carry)
        return dummy.next
